### PR TITLE
Dockerize backend development environments

### PR DESCRIPTION
**Pull request by j-ngo828**
## What does this PR do?

- Dockerize backend development environment
- Add `debugpy` debugger

## Pre-requisites

Install Docker Desktop: https://www.docker.com/products/docker-desktop/

### Setup

Ensure you have Docker Desktop installed and opened. Then run

```bash
docker compose up -d --build
```

This is a sample `.env` file I used to test:

```
DB_TYPE="postgres"
DB_USER="admin"
DB_PASS="abc123"
DB_HOST="db"
DB_PORT=5432
DB_NAME="studyhub"
FLASK_ENV=development
```

Note the `DB_PORT` environment variable, **do not** use `PORT` since deployment service like Fly.io will set a PORT environment variable for the deployed server

Also notice how I don't have Twilio key, thus I wasn't able to properly test the user authentication flow with Postman.

### Verify

- Use Postman, ensure the user authentication flow, user profile CRUD operations, and OAuth still works

Note that you need to add `docker compose exec backend` before the command you want to run. For example, to run `testRegistration.py`:

```bash
docker compose exec backend python Test_RestAPI/testRegistration.py
```

Jack's note: running the above command does not work for some reason. It seems there needs to be additional setup for using the Request library in a container. Try using Postman instead with the base url: `http://0.0.0.0:8080`.

Try using the debugger with VSCode by creating a `.vscode/launch.json` file, if not already exists. Then add the following:

```json
{
  "version": "0.2.0",

  // .... Other settings

  "configurations": [
    {
      "name": "Python: Remote Attach",
      "type": "python",
      "request": "attach",
      "connect": {
        "host": "localhost",
        "port": 5678
      },
      "pathMappings": [
        {
          "localRoot": "${workspaceFolder}",
          "remoteRoot": "."
        }
      ],
      "justMyCode": true
    }
  ]
}
```

Then try setting a breakpoint in one of the router handler, send a request with Postman to that route. Ensure the breakpoint is hit and you can print things out in the VSCode debug console

### CODE REVIEW

File: requirements.old.txt
Comment (j-ngo828) [inline requirements.old.txt:1]: Trimmed `opencv`, `numpy`, and other god knows why it's here packages

File: Pipfile
Comment (j-ngo828) [inline Pipfile:1]: I cannot guarantee this file contains the exact dependencies needed. In other words, it might still contains excessive dependencies. But this should be fine, we can investigate ways to optimize this later.

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+websocket = "==0.2.1"
+werkzeug = "==1.0.1"
+"zope.event" = "==4.6"
+"zope.interface" = "==6.0"
+flask-login = "==0.4.1"
+oauthlib = "==3.0.1"
+pyopenssl = "==19.0.0"
+
+[dev-packages]
+
```
Comment (j-ngo828) [inline Pipfile:92]: Libraries for formatting and linting, as well as testing libraries should be installed as dev dependencies

For now, this is fine. But if you plan to setup CI/CD pipeline then we will need to add more things here

File: Pipfile
Comment (Kenttrann2302) [reply Pipfile:1]: Are all the docker containers set up?

Comment (j-ngo828) [reply Pipfile:1]: Nope, ensure this works first. Then Docker

Comment (Kenttrann2302) [reply Pipfile:1]: Do you want to just merge it in?

Comment (j-ngo828) [reply Pipfile:1]: Nope, this is in scope of the Docker issue. I am testing the container right now

Comment (Kenttrann2302) [reply Pipfile:1]: Okay let me know

**Comment by j-ngo828** [comment]
This doesn't work at all. Me gonna sleep and dream about Tesla interview

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+services:
+  backend:
+    build: .
+    depends_on:
+      db:
+        condition: service_healthy
+    ports:
+      - 8080:5000
```
Comment (j-ngo828) [inline docker-compose.yml:8]: Port 5000 on my machine was being used. Thus I have to map to port 8080 here.

You can try port 5000 on your machine and see what happens

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,24 @@
+FROM python:3.11-slim-buster
```
Comment (j-ngo828) [inline Dockerfile:1]: Using this did slim down the image size quite significantly. But not as slim as an `alpine` image.

I tried setting up `alpine` image but ran into several issues trying to install the dependencies. Let's try setting up that image sometimes in the future for deployment. For now, this is good enough

File: get_env.py
Comment (j-ngo828) [inline get_env.py:1]: This is a benefit of using one source of truth right here.

Notice how I only needed to make changes in exactly 2 places, the `.env` file, and this file that loads the environment variables

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      - 5678:5678
+    volumes:
+      - .:/app
+
+  db:
```
Comment (j-ngo828) [inline docker-compose.yml:17]: One thing to call out is that if you want to run SQL commands, do it in this `db` container

Comment (Kenttrann2302) [reply docker-compose.yml:17]: Okay, i will do the code review tomorrow to see how you set up the Docker containers.

Comment (Kenttrann2302) [reply docker-compose.yml:17]: Actually, everything LGTM. Can we merge this in, can wait to test out our docker containers!

**Review by Kenttrann2302 — APPROVED** [review]
LGTM!

File: docker-compose.yml (hunk 4 above)
Comment (Kenttrann2302) [reply docker-compose.yml:17]: fatal: 'jack/task/dockerize-development-environment' is not a commit and a branch 'task/dockerize-development-environment' cannot be created from it

I don't know why I get this error when I checkout to your branch after I fork your local repo.

Comment (j-ngo828) [reply docker-compose.yml:17]: oh oops, forgot one important command, check the updated comment: https://github.com/Kenttrann2302/StudyHub-Backend/pull/22#discussion_r1183095626

Comment (Kenttrann2302) [reply docker-compose.yml:17]: @j-ngo828 oh no, i think we ran in the same error again with the depencies: 

2023-05-02 18:38:20 sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "127.0.0.1", port 5432 failed: Connection refused
2023-05-02 18:38:20     Is the server running on that host and accepting TCP/IP connections?

This error message from Docker Desktop

Comment (Kenttrann2302) [reply docker-compose.yml:17]: Hang on, I'm so confused, i thought the host and post are already correct?

Comment (j-ngo828) [reply docker-compose.yml:17]: Actually, tell me how did you ran into that issue? Give me a reproduction step

Did the error log occurred after you ran `docker compose up -d --build`?

I just built a fresh image and ran the container on my machine and all dependencies works fine.

Comment (j-ngo828) [reply docker-compose.yml:17]: > Hang on, I'm so confused, i thought the host and post are already correct?

Yup all host and ports are correct

Comment (Kenttrann2302) [reply docker-compose.yml:17]: Yes i did do this: docker compose up -d --build

And I went on docker container to start both the db and the server, the db runs fine but the server exited after 2 seconds and output the log like I showed above.

Comment (j-ngo828) [reply docker-compose.yml:17]: It might be because the `db` image pulled is not specific enough. Give me a sec, I will need to update the `docker-compose.yml` file to pull a more specific `postgres` image

Can you check your `.env` to make sure the port is updated. How about changing the `DB_HOST` to `0.0.0.0` and `DB_PORT` to `5432` in your `.env` file

Comment (Kenttrann2302) [reply docker-compose.yml:17]: I still got the same error when changing the host: sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "0.0.0.0", port 5432 failed: Connection refused
2023-05-02 19:06:25     Is the server running on that host and accepting TCP/IP connections?

Comment (j-ngo828) [reply docker-compose.yml:17]: Can you add the following line of code just before [line 49 in `app.py`](https://github.com/Kenttrann2302/StudyHub-Backend/pull/22/files#diff-568470d013cd12e4f388206520da39ab9a4e4c3c6b95846cbc281abc1ba3c959R49)
```python
raise ValueError(f"{database_type}://{database_username}:{database_password}@{database_host}:{database_port}/{database_name}")
```

Then rerun the docker compose command again and see the value of the ValueError to check if it's correct, reply to this comment with the value of the connection url passed into the ValueError as well. I'm thinking it has to do with incorrect connection url. Check to make sure your `.env` has `DB_PORT` instead of `PORT` too

Comment (Kenttrann2302) [reply docker-compose.yml:17]: Is it okay, since it includes my username and password for my psql server?

Comment (j-ngo828) [reply docker-compose.yml:17]: Hm then let's use the username and password in the `.env` file I have above then.

Actually just copy all the variables I have in the `.env` in the topmost comment in the PR (be sure to save a copy of your `.env` somewhere on your laptop before making this change) to your `.env`, make sure to not have any duplicate variables then retries.

Comment (Kenttrann2302) [reply docker-compose.yml:17]: Okay, I'm feeling very strange that your .env file works, but not for mine???

Comment (j-ngo828) [reply docker-compose.yml:17]: Thank god it works. Wanna hop on a Discord call now and try to reproduce the issue you ran into?

Comment (j-ngo828) [reply docker-compose.yml:17]: sure, no problem

**Comment by j-ngo828** [comment]
#23 should be in scope of this one. In the case there is code change needed it shouldn't blow up the scope of this current PR.

TLDR: to run script in the backend container. Run:

```bash
docker compose exec backend <commands/script to execute here>
```

For example,

```bash
docker compose exec backend flask migrate
```

See: https://docs.docker.com/engine/reference/commandline/compose_exec/

@Kenttrann2302

**Comment by Kenttrann2302** [comment]
@j-ngo828 , I have set up all the POSTMAN API collection and they seem to work very fine with the Docker. Do you want me to push the code and merge this in, since we have one more rest api to go.

**Comment by j-ngo828** [comment]
> @j-ngo828 , I have set up all the POSTMAN API collection and they seem to work very fine with the Docker. Do you want me to push the code and merge this in, since we have one more rest api to go.

Yup please squash and merge it in. Glad it worked
